noir-projects/aztec-nr/aztec/src/utils/bytes.nr

```diff
@@ -0,0 +1,206 @@
+// Converts the input bytes into an array of fields. A Field is ~254 bits meaning that each field can store 31 bytes.
+// This implies that M = ceil(N / 31).
+pub fn bytes_to_fields<let N: u32, let M: u32>(input: [u8; N]) -> [Field; M] {
+    assert(N <= 31 * M, "Bytes do not fit into fields");
+    let mut dst = [0; M];
+
+    for field_index in 0..M {
+        let mut field_value = 0;
+
+        for i in 0..31 {
+            let byte_index = field_index * 31 + i;
+            if byte_index < N {
+                // Shift the existing value left by 8 bits and add the new byte
+                field_value = field_value * 256 + input[byte_index] as Field;
+            }
+        }
+
+        dst[field_index] = field_value;
+    }
+
+    dst
+}
+
+// Converts an input array of fields into bytes. Each field of input has to contain only 31 bytes.
+// TODO(#8618): Optimize for public use.
+pub fn fields_to_bytes<let N: u32, let M: u32>(input: [Field; M]) -> [u8; N] {
+    let mut dst = [0; N];
```

Comment:
We're not constraining or bounding `N` in any way: I can just pass a low `N` value and this function will return the wrong result. Given that we do not have dynamic arrays, we'll either have to guess the output size (which seems bad), return it as a second result value and have padding (seems bad too since the return array will have enlarged size), or maybe store it alongside the field array (so that when going from u8 to field we keep track of the original length). I imagine we'd otherwise have a lot of trouble actually using this function if not.